### PR TITLE
Generalize moving of identical code from if/select arms

### PR DESCRIPTION
**Pull request by kripken**
* Effects are fine in the moved code, if we are doing so on an `if`
  (which runs just one arm anyhow).
* Allow `unreachable`, which lets us hoist `return`s for example.
* Allow `none,` which lets us hoist `drop` and `call` for example. For
  this we also need to be careful with subtyping, as at least `drop`
  is polymorphic, so the child types may not have an LUB (see
  example in code).

Adds a small ShallowEffectAnalyzer child of EffectAnalyzer that
calls `visit` to just do a shallow analysis (instead of `walk` which
walks the children).

### CODE REVIEW

File: src/passes/OptimizeInstructions.cpp [hunk 1]
```diff
@@ -2860,27 +2856,42 @@ struct OptimizeInstructions
...
       while (1) {
-        // TODO: handle none as well as concrete types (can pull a drop out, for
-        //       example)
-        // TODO: consider the case with more children than 1
-        if (curr->ifTrue->type.isConcrete() &&
-            !Properties::isControlFlowStructure(curr->ifTrue) &&
+        // Ignore control flow structures (which are handled in MergeBlocks).
+        if (!Properties::isControlFlowStructure(curr->ifTrue) &&
             ExpressionAnalyzer::shallowEqual(curr->ifTrue, curr->ifFalse)) {
+          // TODO: consider the case with more children than 1
```
Comment (tlively) [inline src/passes/OptimizeInstructions.cpp:2862]:
```suggestion
          // TODO: consider the case with more than one child.
```

File: test/lit/passes/optimize-instructions.wast [hunk 2]
```diff
@@ -12013,4 +12024,108 @@
...
+  ;; CHECK-NEXT:     (local.get $y)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $ternary-identical-arms-br_if-same (param $x i32) (param $y i32) (param $z i32)
+    (block $block
+      (if
+        (local.get $z)
+        ;; two br_ifs with different targets are shallowly identical
```
Comment (tlively) [inline test/lit/passes/optimize-instructions.wast:12042]:
```suggestion
        ;; two br_ifs with the same target are shallowly identical
```

File: test/lit/passes/optimize-instructions.wast [hunk 3]
```diff
@@ -12013,4 +12024,108 @@
       )
     )
   )
+  ;; CHECK:      (func $ternary-identical-arms-br_if-same (param $x i32) (param $y i32) (param $z i32)
+  ;; CHECK-NEXT:  (block $block
+  ;; CHECK-NEXT:   (br_if $block
+  ;; CHECK-NEXT:    (if (result i32)
+  ;; CHECK-NEXT:     (local.get $z)
+  ;; CHECK-NEXT:     (local.get $x)
+  ;; CHECK-NEXT:     (local.get $y)
```
Comment (tlively) [inline test/lit/passes/optimize-instructions.wast:12033]: It looks like we could further simplify `if`s into `select`s whenever the only side effects are reading.

**Review by tlively — APPROVED** [review]
Nice 👍

File: test/lit/passes/optimize-instructions.wast (hunk 3 above)
Comment (kripken) [reply test/lit/passes/optimize-instructions.wast:12033]: I believe we already do that. A read is not a side effect, and RemoveUnusedBrs has a `selectify` method to turn relevant ifs into selects. It has some heuristics based on how expensive executing both sides,

https://github.com/WebAssembly/binaryen/blob/cb776da297a845e15171405b7f7518619122d7aa/src/passes/RemoveUnusedBrs.cpp#L1057-L1063

**Pull request merged**